templates/service-worker.js

```diff
@@ -94,3 +94,123 @@ self.addEventListener('fetch', event => {
     return fetch(req)
   })())
 })
+
+self.addEventListener('push', event => {
+  event.waitUntil((async () => {
+    const data = event.data.json()
+    const { origin } = new URL(data.icon)
+
+    try {
+      const notification = await get(`${origin}/api/v1/notifications/${data.notification_id}`, {
+        'Authorization': `Bearer ${data.access_token}`,
+      }, { timeout: 2000 })
+
+      await showRichNotification(data, notification)
+    } catch (e) {
+      await showSimpleNotification(data)
+    }
+  })())
+})
+
+async function showSimpleNotification(data) {
+  await self.registration.showNotification(data.title, {
+    icon: data.icon,
+    body: data.body
+  })
+}
+
+async function showRichNotification(data, notification) {
+  switch (notification.type) {
```

Comment:
These are very similar right now but I prefer to handle them separately since they may diverge in the future (it is also clearer this way).

Reply:
:+1: